packages/dds/map/src/mapIncremental.ts

```diff
@@ -0,0 +1,466 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	ISequencedDocumentMessage,
+	MessageType,
+	SummaryType,
+} from "@fluidframework/protocol-definitions";
+import {
+	IChannelAttributes,
+	IFluidDataStoreRuntime,
+	IChannelStorageService,
+	IChannelServices,
+	IChannelFactory,
+} from "@fluidframework/datastore-definitions";
+import {
+	IIncrementalContext,
+	ISummaryTreeWithStats,
+	ITelemetryContext,
+} from "@fluidframework/runtime-definitions";
+import { readAndParse } from "@fluidframework/driver-utils";
+import { IFluidSerializer, SharedObject } from "@fluidframework/shared-object-base";
+import { SummaryTreeBuilder } from "@fluidframework/runtime-utils";
+import { ISharedMap, ISharedMapEvents } from "./interfaces";
+import { IMapDataObjectSerializable, IMapOperation, MapKernel } from "./mapKernel";
+import { pkgVersion } from "./packageVersion";
+
+interface IMapSerializationFormat {
+	blobs?: string[];
+	content: IMapDataObjectSerializable;
+}
+
+const snapshotFileName = "header";
+
+function addBlobOrHandle(
+	parentPath: string,
+	blobName: string,
+	previousBlob: string,
+	currentBlob: string,
+	builder: SummaryTreeBuilder,
+) {
+	if (currentBlob === previousBlob) {
```

Comment:
This will work, but this is very expensive - DDS need to generate whole summary in memory and do this diff to learn if it's generating same blob. This is much more efficient to do not at that level, but as an adapter over storage interface - we can build generic blob de-dupper at whole runtime layer that would de-dup blobs even across DDSs.

What we really want is (in priority order)
1. ability for DDS not to do this expensive recalc to figure out that nothing stored in a blob changed and thus it can be reused
2. Ideally, not even to have that content in memory (if DDS is structured that way). I.e. imaging that map's content is bigger than memory, but only 1% of it ever is used. We want to create infrastructure that allows building such DDSs.

Reply:
BTW, if we were to go that approach, or approach of adapter, we would not keep in memory these blobs. We would keep only hashes of blobs and would compare hashes - that way no need to duplicate memory usage.